src/combinators/mod.rs

```diff
@@ -1,35 +1,46 @@
 use gen::GenError;
 use lib::std::io::Write;
+use lib::std::io;
 
-pub type GenResult<I> = Result<(I, usize), GenError>;
+pub type GenResult<I> = Result<I, GenError>;
 
 pub trait SerializeFn<I>: Fn(I) -> GenResult<I> {}
 
 impl<I, F: Fn(I) -> GenResult<I>> SerializeFn<I> for F {}
 
-pub trait GenChain<I, S: SerializeFn<I>> {
-    fn chain(self, s: &S) -> GenResult<I>;
+pub trait Skip {
+    fn skip(self, s: usize) -> Result<Self, GenError> where Self: Sized;
 }
 
-impl<I, S: SerializeFn<I>> GenChain<I, S> for GenResult<I> {
-    fn chain(self, s: &S) -> GenResult<I> {
-        self.and_then(|(buf, len)| s(buf).map(|(buf, len2)| (buf, len + len2)))
+pub struct WriteCounter<W: Write>(W, usize);
```

Comment:
You should probably make inner fields public

Reply:
That would allow to break the invariants of the type. You could replace the `W` then, change the `position`, etc.

What use-case do you see for making the `W` or `position` public, in addition to the functions there are there already?

Reply:
I want to be able to wrap my Write only for one call and get it back

Reply:
How about adding an `fn into_inner(self) -> W` then? That would allow you to do that without allowing to break the invariants of the type from the outside

Or maybe `fn into_inner(self) -> (W, usize)` but that seems unnecessary as there's also the `fn position(&self) -> usize`.

I'm also wondering if this should be `u64` instead of `usize`. I guess it should, to allow >4GB writes on 32 bit platforms.

Reply:
Well, usize matches what Write does, and u32 is slowly dying...

Into inner returning the tuple would be great!

Reply:
> Well, usize matches what Write does, and u32 is slowly dying...

For a single write, `usize` is also the limit (as it's the maximum size of a single slice). But writing multiple slices can easily get over that limit.

If you check e.g. the `std::io::Seek` trait, it works with `u64` for that very reason.

Reply:
I’m fine with both.
Let’s go with u64 then